superchain-backend.md

```diff
@@ -0,0 +1,457 @@
+    OP Mainnet at 4M logs per day. -> About 93 logs per block.
+- Other L2 chains:
+  - Base at 11M logs per day.
+  - Zora at 0.5-2M logs per day.
+  - Arbitrum at 8M logs per day.
+  - Ethereum L1 at 3M logs per day.
+
+The history of these log events sums up to:
+- OP Mainnet history (from Dune query) = 1.9B logs.
+- Size of just 32 bytes per log = ~ 60 GB.
+- Size of receipts in geth (from recent `db inspect`) = 80+ GB.
+
+Caching a day worth of event log hashes is ~130 MB for OP Mainnet.
+The worst-case, if a chain is attacked with unbalanced log-event usage, has significantly more logs than that.
+This would amount to `13k / 93 = ~ 140` times as many log-events,
+or about 18 GB of log-event hashes for a full-day full-capacity attack on one standard-throughput chain.
+
+A standalone-service that prefetches the logs should be able to support the above:
+- A trivial amount of memory for the last ~1 hour of block-hashes.
+- An in-memory cache for recent ~1 hour of log-events, expected to be the most commonly consumed,
+  as well as cover all recent executing messages that are yet to be verified, or form a dependency that may reorg out.
+- Disk storage, around ~65 GB per L2 chain of OP-Mainnet scale and history-size,
+  for an append-only log of block-hashes and event-log hashes per chain.
+
+Note that there does not have to be any indexing of hashes back to metadata,
+as the executing-messages include the block-number and event-log index, for fast lookups.
+And a binary-search over the append-only log can find the data quickly:
+there is no need for a more free key-value database.
+Optionally the hashes may be truncated to ~20 bytes, like ethereum addresses,
+to save on storage while still being reasonably cryptographically secure.
+
+
+## Responsibilities
+
+The responsibilities of the superchain backend are to:
+- (pre-)fetch executing messages of each chain.
+- Track cross-unsafe-head of each chain.
+- Track unsafe head of each chain, to handle reorgs.
+- Resolve block and message safety.
+- And possible future extensions, post-launch:
+  - Track shreds of each chain (inputs to intra-block message safety).
+  - Resolve block safety incl. cycles (intra-block message safety).
+
+## Implementation ideas
+
+### Verify as hashes, not messages
+
+Log-events can be huge, their size is variable.
+Importantly, an executing message already encodes what is being executed.
+Verifying should not require retrieval of the full initiating message.
+
+And there is no "partially valid":
+the log-event contract-origin address and log-event topics and data can be hashed into one commitment.
+
+Since initiating messages and executing messages are both log events,
+and assuming intra-block messaging is allowed by the protocol, we can:
+
+- use an executing message as initiating message
+- execute the same message
+- have a loop of executing messages between chains
+
+We should version the hash, and type it, such that we can:
```

Comment:
Do you suggest making this hash an `indexed` field, so that it can be used to identify a log?

Reply:
For now it's an offchain detail. We can include it onchain, but it would add overhead, and we can't expect it from initiating messages anyway. The event-hashing here applies to both executing and initiating messages, so there is less data to retain and process offchain.